rust/datafusion/src/physical_optimizer/repartition.rs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Repartition optimizer that introduces repartition nodes to increase the level of parallism available
+use std::sync::Arc;
+
+use super::optimizer::PhysicalOptimizerRule;
+use crate::physical_plan::{repartition::RepartitionExec, ExecutionPlan};
+use crate::physical_plan::{Distribution, Partitioning::*};
+use crate::{error::Result, execution::context::ExecutionConfig};
+
+/// Optimizer that introduces repartition to introduce more parallelism in the plan
+pub struct Repartition {}
+
+impl Repartition {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+fn optimize_concurrency(
+    concurrency: usize,
+    requires_single_partition: bool,
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Arc<dyn ExecutionPlan>> {
+    // Recurse into children bottom-up (added nodes should be as deep as possible)
+
+    let new_plan = if plan.children().is_empty() {
+        // leaf node - don't replace children
```

Comment:
it seems like `new_with_children` could handle the case of zero children as well, FWIW